File: 2023/UICPC Regional/Round #8/solutions/SuffixArrayREconstruction/solutions/tobi.cpp
// @EXPECTED_RESULTS@: CORRECT

#include <stdio.h>
#include <string.h>

#define FOR(i,a,b) for (int i = (a); i < (b); i++)

int main(int argc, char **argv) {
	int numCases;
	scanf("%d", &numCases);

	while (numCases --> 0) {
		int len, numSuff, pos;
		char buf[10005], suff[10005];

		scanf("%d%d", &len, &numSuff);
		FOR(i, 0, len) buf[i] = '*';
		buf[len] = '\0';

		bool possible = true;
		FOR(i, 0, numSuff) {
			scanf("%d%s", &pos, suff);
			pos--;
			int suffLen = strlen(suff);
			FOR(j, 0, suffLen) {
				if (suff[j] == '*') {
					break;
				} else if (buf[j + pos] == '*' || buf[j + pos] == suff[j]) {
					buf[j + pos] = suff[j];
				} else {
					possible = false;
					break;
				}
			}
			FOR(j, 0, suffLen) {
				if (suff[suffLen - j - 1] == '*') {
					break;
				} else if (buf[len - j - 1] == '*' || buf[len - j - 1] == suff[suffLen - j - 1]) {
					buf[len - j - 1] = suff[suffLen - j - 1];
				} else {
					possible = false;
					break;
				}
			}
		}
		FOR(i, 0, len) {
			if (buf[i] == '*') {
				possible = false;
			}
		}
		if (possible) {
			printf("%s\n", buf);
		} else {
			puts("IMPOSSIBLE");
		}
	}

	return 0;
}

File: 2023/UICPC Regional/Round #8/solutions/SuffixArrayREconstruction/solutions/tobi.java
// @EXPECTED_RESULTS@: CORRECT

import java.io.*;
import java.util.*;

public class tobi {
	public static void main(String args[]) {
		Scanner sc = new Scanner(System.in);

		int numCases = sc.nextInt();

		while (numCases --> 0) {
			int len, numSuff, pos;
			String buf, suff;

			len = sc.nextInt();
			numSuff = sc.nextInt();
			char bufChars[] = new char[len];
			for (int i = 0; i < len; i++) {
				bufChars[i] = '*';
			}

			boolean possible = true;
			for (int i = 0; i < numSuff; i++) {
				pos = sc.nextInt() - 1;
				suff = sc.next();
				char suffChars[] = suff.toCharArray();
				int suffLen = suff.length();
				for (int j = 0; j < suffLen; j++) {
					if (suffChars[j] == '*') {
						break;
					} else if (bufChars[j + pos] == '*' || bufChars[j + pos] == suffChars[j]) {
						bufChars[j + pos] = suffChars[j];
					} else {
						possible = false;
						break;
					}
				}
				for (int j = 0; j < suffLen; j++) {
					if (suffChars[suffLen - j - 1] == '*') {
						break;
					} else if (bufChars[len - j - 1] == '*' || bufChars[len - j - 1] == suffChars[suffLen - j - 1]) {
						bufChars[len - j - 1] = suffChars[suffLen - j - 1];
					} else {
						possible = false;
						break;
					}
				}
			}
			for (int i = 0; i < len; i++) {
				if (bufChars[i] == '*') {
					possible = false;
				}
			}
			if (possible) {
				System.out.println(bufChars);
			} else {
				System.out.println("IMPOSSIBLE");
			}
		}
	}
}
